internal/pkg/agent/application/upgrade/step_download.go

```diff
@@ -81,10 +81,13 @@ func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI stri
 		return "", errors.New(err, "initiating verifier")
 	}
 
-	if err := verifier.Verify(agentArtifact, parsedVersion.VersionWithPrerelease(), skipDefaultPgp, pgpBytes...); err != nil {
+	success, err := verifier.Verify(agentArtifact, parsedVersion.VersionWithPrerelease(), skipDefaultPgp, pgpBytes...)
+	if !success {
 		return "", errors.New(err, "failed verification of agent binary")
 	}
-
+	if err != nil {
```

Comment:
This isn't really the standard way to do this, if the operation succeeded `err` should be `nil`. 

Now that I've seen this approach, I think a better solution would be to have the composed verifier log each verifier that failed but return only a `nil` error if at least one of them succeeded.

If you wanted to keep doing it this way, you would want to return the details of what happened as a separate structure so it can be logged at this level but I think it will be much simpler to handle the logging directly in this block of the composed verifier

Something like the following should handle most of this, it already builds up a joined error of each failure and only returns it if all of the verifiers fail. We just need to add a log statement and a `Name` method to the verifier interface so you can log which one failed.

```diff
--- a/internal/pkg/agent/application/upgrade/artifact/download/composed/verifier.go
+++ b/internal/pkg/agent/application/upgrade/artifact/download/composed/verifier.go
@@ -5,7 +5,7 @@
 package composed

 import (
-       "github.com/hashicorp/go-multierror"
+       "errors"

        "github.com/elastic/elastic-agent/internal/pkg/agent/application/upgrade/artifact"
        "github.com/elastic/elastic-agent/internal/pkg/agent/application/upgrade/artifact/download"
@@ -43,11 +43,10 @@ func (e *Verifier) Verify(a artifact.Artifact, version string, skipDefaultPgp bo
                        return nil
                }

-               err = multierror.Append(err, e)
+               err = errors.Join(err, e)

                if errors.As(e, &checksumMismatchErr) || errors.As(err, &invalidSignatureErr) {
-                       // Stop verification chain on checksum/signature errors.
-                       break
+                       log.Warningw("Verifier failed!", "verifier", v.Name(), "error", err)
                }
        }
```

Reply:
You would still want the logger here to log the case where the all fail, which should be logged at the `Error` level.

You should be able to pass the logger you have here into the function call.

Reply:
Yes, we need success = "no error".

The trouble you are running into is that you need to log twice, once inside the `composed/verifier.go` file for non-fatal errors that won't be returned (since final success = no error) and once when the verifier is called from the upgrade process for the fatal error where all verifiers failed.